arch/x86/trusty.c

```diff
@@ -96,56 +96,68 @@ _Static_assert(sizeof(struct trusty_startup_param)
 	exec_vmwrite(VMX_GUEST_##SEG_NAME##_ATTR, seg.attr); \
 }
 
-void create_secure_world_ept(struct vm *vm, uint64_t gpa,
-		uint64_t size, uint64_t rebased_gpa)
+static void create_secure_world_ept(struct vm *vm, uint64_t gpa_orig,
+		uint64_t size, uint64_t gpa_rebased)
 {
-	int i = 0;
-	uint64_t pml4e = 0;
-	uint64_t entry = 0;
+	uint64_t nworld_pml4e = 0;
+	uint64_t sworld_pml4e = 0;
 	struct map_params map_params;
-	uint64_t hpa = gpa2hpa(vm, gpa);
+	uint64_t gpa = 0;
+	uint64_t hpa = gpa2hpa(vm, gpa_orig);
+	uint64_t table_present = (IA32E_EPT_R_BIT |
+				IA32E_EPT_W_BIT |
+				IA32E_EPT_X_BIT);
+	void *sub_table_addr = NULL;
 	struct vm *vm0 = get_vm_from_vmid(0);
 
-	/* Create secure world eptp */
-	if (vm->sworld_control.sworld_enabled && !vm->arch_vm.sworld_eptp)
-		vm->arch_vm.sworld_eptp = alloc_paging_struct();
+	if (!vm->sworld_control.sworld_enabled
+			|| vm->arch_vm.sworld_eptp != NULL) {
+		pr_err("Sworld is not enabled or Sworld eptp is not NULL");
+		return;
+	}
 
 	map_params.page_table_type = PTT_EPT;
 	map_params.pml4_inverted = vm->arch_vm.m2p;
 
-	/* unmap gpa~gpa+size from guest ept mapping */
+	/* Unmap gpa_orig~gpa_orig+size from guest normal world ept mapping */
 	map_params.pml4_base = vm->arch_vm.nworld_eptp;
-	unmap_mem(&map_params, (void *)hpa, (void *)gpa, size, 0);
+	unmap_mem(&map_params, (void *)hpa, (void *)gpa_orig, size, 0);
 
 	/* Copy PDPT entries from Normal world to Secure world
 	 * Secure world can access Normal World's memory,
 	 * but Normal World can not access Secure World's memory.
-	 * The PML4/PDPT for Secure world are separated from Normal World.
-	 * PD/PT are shared in both Secure world's EPT and Normal World's EPT
+	 * The PML4/PDPT for Secure world are separated from
+	 * Normal World.PD/PT are shared in both Secure world's EPT
+	 * and Normal World's EPT
 	 */
-	for (i = 0; i < IA32E_NUM_ENTRIES; i++) {
-		pml4e = MEM_READ64(vm->arch_vm.nworld_eptp);
-		entry = MEM_READ64((pml4e & IA32E_REF_MASK)
-				+ (i * IA32E_COMM_ENTRY_SIZE));
-		pml4e = MEM_READ64(vm->arch_vm.sworld_eptp);
-		MEM_WRITE64((pml4e & IA32E_REF_MASK)
-				+ (i * IA32E_COMM_ENTRY_SIZE),
-				entry);
-	}
+	vm->arch_vm.sworld_eptp = alloc_paging_struct();
+
+	/* The trusty memory is remapped to guest physical address
+	 * of 511G to 511GB + 16MB
```

Comment:
how about using corresponding macros for 511GB and 16MB? (16MB will potentially be changed in future)

Reply:
here the map is according to the input parameters(rebased_gpa and size), i will update the comment